### PR TITLE
Fix Rollback NRTMv4 Client

### PR DESCRIPTION
**Pull request by MiguelAHM**
- When using "parallel()" each record will use one transaction, so rollback is not possible because each record is being committed.
So I have to refactor to use ExecutorService and manually manage the transactions. I am creating a transaction commit per batch.
So there could be an issue with the third batch, when two previous batch have been already committed. So in this case the third batch will rollback and there will be a truncate tables. So the database will be clean.

- For rest of the situations I relay in the transaction. If deltas fails the transaction will take place
- Refactor to persist the version at the end of the import. I am using a different database, so a different datasource for version. So I cannot rollback if I add the snapshot before processing the records and there is a problem with processing the records (with the current setup, transaction chains are no set). So I order the code for using first one datasource (doing the import) and the another (add version) and not in the other way around

### CODE REVIEW

File: whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/SnapshotMirrorImporter.java [hunk 1]
```diff
@@ -145,31 +145,59 @@ public void run() {
...
+                    final Map.Entry<RpslObject, RpslObjectUpdateInfo> persistedRecord = nrtm4ClientRepository.processSnapshotRecord(mirrorRpslObject);
+                    nrtm4ClientRepository.createIndexes(persistedRecord.getKey(), persistedRecord.getValue());
+                    processedCount.incrementAndGet();
+                } catch (Exception e) {
+                    nrtmClientUpdateTransaction.rollback(transactionStatus);
+                    LOGGER.error("Unable to process record {}", record, e);
+                    throw new IllegalStateException(e);
+                }
+            });
+        } catch (Exception e) {
```
Comment (eshryane) [inline whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/SnapshotMirrorImporter.java:196]: I don't think the outside try/catch is useful, unless you expect `submit()` to fail. The record is processed asynchronously. You could return the `Future` and check in the loop whether the job failed, and then stop submitting jobs if any failed.

Comment (eshryane) [reply whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/SnapshotMirrorImporter.java:196]: Or use a global `AtomicBoolean` to flag if any job failed and if so, stop all jobs immediately.

Comment (MiguelAHM) [reply whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/SnapshotMirrorImporter.java]: I go with future for handling errors. So I check for the errors

File: whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/SnapshotMirrorImporter.java [hunk 2]
```diff
@@ -62,32 +86,51 @@ public void doImport(final String source,
...
         }
 
+        final AtomicInteger snapshotVersion = new AtomicInteger(0);
         final AtomicInteger processedCount = new AtomicInteger(0);
+
+        persisSnapshot(source, payload, sessionId, snapshotVersion, processedCount);
+        persistSnapshotVersion(source, snapshotVersion.get(), sessionId);
+
+        stopwatch.stop();
+        LOGGER.info("Loading snapshot file took {} for source {} and added {} records", stopwatch.elapsed().toMillis(), source, processedCount.get());
```
Comment (eshryane) [inline whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/SnapshotMirrorImporter.java:96]: No need to convert snapshot to milliseconds, the `toString()` will always print something readable.

File: whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/SnapshotMirrorImporter.java [hunk 3]
```diff
@@ -33,16 +50,23 @@ public class SnapshotMirrorImporter extends AbstractMirrorImporter {
...
                                   final Nrtm4ClientInfoRepository nrtm4ClientInfoMirrorDao,
-                                  final Nrtm4ClientRepository nrtm4ClientRepository) {
+                                  final Nrtm4ClientRepository nrtm4ClientRepository,
+                                  @Qualifier(NRTM_CLIENT_UPDATE_TRANSACTION) final PlatformTransactionManager transactionManagerNrtmClientUpdate) {
         super(nrtm4ClientInfoMirrorDao, nrtm4ClientRepository);
         this.nrtmRestClient = nrtmRestClient;
+        this.nrtmClientUpdateTransaction =  transactionManagerNrtmClientUpdate;
 
+        final int numThreads = Runtime.getRuntime().availableProcessors();
+        this.numberOfThreads = (int)Math.ceil(numThreads/4.0);
```
Comment (eshryane) [inline whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/SnapshotMirrorImporter.java:67]: No need for a float, you can also use `Math.min(1, numThreads/4)`

File: whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/config/NrtmClientTransactionConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+import javax.sql.DataSource;
+
+
+@Configuration
+@Conditional(Nrtm4ClientCondition.class)
+@EnableTransactionManagement(mode = AdviceMode.ASPECTJ)
+public class NrtmClientTransactionConfiguration {
+
+    public static final String NRTM_CLIENT_UPDATE_TRANSACTION = "nrtm-client-update-transaction-manager";
```
Comment (eshryane) [inline whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/config/NrtmClientTransactionConfiguration.java:21]: No need to define fields if value is only used once

File: whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/dao/Nrtm4ClientInfoRepository.java [hunk 5]
```diff
@@ -64,7 +68,7 @@ SELECT id, source, MAX(version), session_id, type, hostname, created
     @Nullable
     public NrtmClientVersionInfo getNrtmLastVersionInfoForDeltasPerSource(final String source){
         try {
-            return jdbcMasterTemplate.queryForObject("""
+            return jdbcSlaveTemplate.queryForObject("""
```
Comment (eshryane) [inline whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/dao/Nrtm4ClientInfoRepository.java:71]: Why are you using the *slave* datasource here when elsewhere you're using the master datasource during import ? Is it possible that the slave datasource will be out-of-sync ?

File: whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/SnapshotMirrorImporter.java [hunk 6]
```diff
@@ -105,20 +148,72 @@ public void run() {
...
+                    processedCount.incrementAndGet();
+                } catch (Exception e) {
+                    nrtmClientUpdateTransaction.rollback(transactionStatus);
+                    LOGGER.error("Unable to process record {}", record, e);
+                    throw new IllegalStateException(e);
+                }
+            });
     }
 
     public static RpslObject getPlaceholderPersonObject() {
```
Comment (eshryane) [inline whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/SnapshotMirrorImporter.java:219]: This method has been deleted in a separate PR correct?

File: whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/dao/Nrtm4ClientInfoRepository.java (hunk 5 above)
Comment (MiguelAHM) [reply whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/dao/Nrtm4ClientInfoRepository.java:71]: yes, I will use master

File: whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/SnapshotMirrorImporter.java [hunk 7]
```diff
@@ -1,29 +1,46 @@
...
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 
+import static net.ripe.db.nrtm4.client.config.NrtmClientTransactionConfiguration.NRTM_CLIENT_UPDATE_TRANSACTION;
+
 
 @Service
 @Conditional(Nrtm4ClientCondition.class)
```
[the remark this answers is not in the document]
Comment (MiguelAHM) [reply whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/SnapshotMirrorImporter.java:46]: yes, deltas and snapshot
and there is common test between them

File: whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/config/NrtmClientTransactionConfiguration.java (hunk 4 above)
Comment (MiguelAHM) [reply whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/config/NrtmClientTransactionConfiguration.java:21]: It is public, it is being in use by repository layer...
Is how TransactionConfiguration works as well

File: whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/SnapshotMirrorImporter.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (eshryane) [reply whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/SnapshotMirrorImporter.java:46]: OK

File: whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/dao/Nrtm4ClientInfoRepository.java (hunk 5 above)
Comment (MiguelAHM) [reply whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/dao/Nrtm4ClientInfoRepository.java]: Done

**Review by eshryane — APPROVED** [review]

**Pull request merged**